kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -286,28 +302,40 @@ public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata me
     }
   }
 
-  public static final TableFeature TYPE_WIDENING_RW_FEATURE =
-      new TypeWideningTableFeature("typeWidening");
-  public static final TableFeature TYPE_WIDENING_PREVIEW_TABLE_FEATURE =
-      new TypeWideningTableFeature("typeWidening-preview");
-
-  private static class TypeWideningTableFeature extends TableFeature.ReaderWriterFeature
-      implements FeatureAutoEnabledByMetadata {
-    TypeWideningTableFeature(String featureName) {
+  /* ---- Start: type widening ---- */
+  // Base class for typeWidening and typeWidening-preview features. Both features are same in terms
+  // of behavior and given the feature is graduated, we will enable the `typeWidening` by default
+  // if the metadata requirements are satisfied and the table doesn't already contain the
+  // `typeWidening-preview` feature.
+  private static class TypeWideningTableFeatureBase extends TableFeature.ReaderWriterFeature {
+    TypeWideningTableFeatureBase(String featureName) {
       super(featureName, /* minReaderVersion = */ 3, /* minWriterVersion = */ 7);
     }
+  }
 
-    @Override
-    public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata metadata) {
-      return TableConfig.TYPE_WIDENING_ENABLED.fromMetadata(metadata);
+  private static class TypeWideningTableFeature extends TypeWideningTableFeatureBase
+      implements FeatureAutoEnabledByMetadata {
+    TypeWideningTableFeature() {
+      super("typeWidening");
     }
 
     @Override
-    public boolean hasKernelWriteSupport(Metadata metadata) {
-      return true;
+    public boolean metadataRequiresFeatureToBeEnabled(Protocol protocol, Metadata metadata) {
+      return TableConfig.TYPE_WIDENING_ENABLED.fromMetadata(metadata)
+          &&
+          // Don't automatically enable the stable feature if the preview feature is already
+          // supported, to
+          // avoid possibly breaking old clients that only support the preview feature.
+          !protocol.supportsFeature(TYPE_WIDENING_RW_PREVIEW_FEATURE);
```

Comment:
Not necessarily for this PR but supportsFeature doesn't read well to me here (I would expect the predicate to be protocol.FeatureEnabled(....)

Reply:
We use the word `support` to indicate the protocol has the feature (i.e. the table has contents that have the feature). Enabled is used to indicate whether the feature is active in operations (i.e.).